### DISCUSSION TITLE
PriorityQueue based on Okasaki's "Purely Functional Data Structures"

### DISCUSSION THREAD

**Discussion by dvberkel** [uwu-tech/Kind#274] [Show and tell]
I have created a [pull-request](https://github.com/uwu-tech/Kind/pull/272) that adds a `PriorityQueue` to base.

A [priority queue](https://en.wikipedia.org/wiki/Priority_queue) is a 

>  an abstract data type similar to a regular queue or stack data structure in which each element additionally has a "priority" associated with it. In a priority queue, an element with high priority is served before an element with low priority.

In the implementation a priority is function `A -> Nat`. A smaller natural number corresponds with a higher priority.

I would love to receive feedback. Specifically on style, implementation and improvements.

**Comment by dvberkel** [comment]
I am looking for some theorems to proof so I can cement my knowledge of Kind's theorem proving

**Reply by rigille** [reply]
Sure here are some definitions for you to try ^^
```js
// Proof element is in queue
PriorityQueue.In<A: Type>(tree: PriorityQueue<A>, a: A): Type
  ?PriorityQueue.In // i suggest you do a case in the underlying tree

// just a shorthand to apply the priority function
PriorityQueue.priority<A: Type>(tree: PriorityQueue<A>, a: A): Nat
  ?PriorityQueue.priority

// function that's guaranteed to return a value, the hypothesis is erased at runtime for efficiency
PriorityQueue.safe_head<A: Type>(queue: PriorityQueue<A>)<Hyp: PriorityQueue.is_empty!(queue) == false>:  A
  ?PriorityQueue.is_empty

// if a queue has a member, then it isn't empty
PriorityQueue.In.not_empty<A: Type>(
  queue: PriorityQueue<A>
  a: A
  Hyp: PriorityQueue.In<A>(queue, a)
): PriorityQueue.is_empty!(queue) == false
  ?PriorityQueue.In.not_empty

// Proof head has highest priority
PriorityQueue.head.ok<A: Type>(
  tree: PriorityQueue<A>
  a: A
  Hyp: PriorityQueue.In!(queue, a)
): (PriorityQueue.priority!(queue, PriorityQueue.safe_head!(queue, PriorityQueue.In.not_empty!(queue, a, Hyp))
    >=? PriorityQueue.priority!(queue, a)) == true
  ?PriorityQueue.head.ok
```
Those are quite a challenge and will likely require lemmas. The main problems you'll face is that expressions start to get big, and the goals lose readability. Also you'll likely need some type annotations to reduce a `case` on expressions, as in https://github.com/uwu-tech/Kind/discussions/247. The good news is that you're using `Nat`s to measure priority and that makes things much simpler!

**Reply by dvberkel** [reply]
Those are great suggestions. I will start working on them.

I can imagine that I will have a lot of questions once I am down in the trenches

**Reply by dvberkel** [reply]
I worked on the following suggestions.

1. priority
2. safe_head

I want to start with the implementation of In. I was wondering on the suggested definition. Is there a similar definition in base that I can look at?

**Reply by rigille** [reply]
Sure, have a look at `User.rigille.List.In`. In your case you'll have three possibilities so it's best if you define an auxiliary type with three nicely named constructors instead of using `Or`.

**Reply by rigille** [reply]
Also be careful with this current limitation of the smart motive: https://github.com/uwu-tech/Kind/issues/278. I hope to fix it soon ^^

**Reply by dvberkel** [reply]
The following is my attempt to work towards the `In` type for queues. It delegates to an `In` type for trees. But the following seems to trip up the Kind type checker:

```kind
PriorityQueue.Tree.In<A: Type>(tree: PriorityQueue.Tree<A>, a: A): Type
case tree {
    empty: Empty
    node:
      ?a
    //   PriorityQueue.Tree.Location!(tree, a)<hypothesis>
}

PriorityQueue.Tree.left_of<A: Type>(tree: PriorityQueue.Tree<A>)<Hyp: PriorityQueue.Tree.is_empty!(tree) == false>: PriorityQueue.Tree<A>
  case tree with Hyp {
      empty:
        let hypothesis = Hyp :: rewrite x in (x == false) with PriorityQueue.Tree.empty_tree_is_empty<A>
        let impossible = Bool.true_neq_false(hypothesis)
        Empty.absurd!(impossible)
      node:
        tree.left
  }!

PriorityQueue.Tree.right_of<A: Type>(tree: PriorityQueue.Tree<A>)<Hyp: PriorityQueue.Tree.is_empty!(tree) == false>: PriorityQueue.Tree<A>
  case tree with Hyp {
      empty:
        let hypothesis = Hyp :: rewrite x in (x == false) with PriorityQueue.Tree.empty_tree_is_empty<A>
        let impossible = Bool.true_neq_false(hypothesis)
        Empty.absurd!(impossible)
      node:
        tree.right
  }!

type PriorityQueue.Tree.Location<A: Type>(tree: PriorityQueue.Tree<A>, a: A)<Hyp: PriorityQueue.Tree.is_empty!(tree) == false> {
  here(prf: PriorityQueue.Tree.peek!(tree)<Hyp> == a)
  left(prf: PriorityQueue.Tree.Location!(PriorityQueue.Tree.left_of!(tree)<Hyp>, a))
  right(prf: PriorityQueue.Tree.Location!(PriorityQueue.Tree.right_of!(tree)<Hyp>, a))
}
```

And I am not sure why. The code can be found in situ in [the branch for the pull-request](https://github.com/fifth-postulate/Kind/tree/priorityqueue)

When I run `kind-scm PriorityQueue.Tree.In` it does not seem to halt. I was expecting to say something clever to make me slap my forehead

**Reply by rigille** [reply]
Congratulations you found a bug in the desugaring process of datatype definitions! It's described in https://github.com/uwu-tech/Kind/issues/281.

That's not what's causing the loop however. There's a missing parameter at the end of, for example, `PriorityQueue.Tree.Location!(PriorityQueue.Tree.left_of!(tree)<Hyp>, a)`. It cannot be filled because left_of tree could be empty.

How about this?
```
type PriorityQueue.Either <A; Type, B: Type, C: Type> {
  left(val: A)
  here(val: B)
  right(val: C)
}
PriorityQueue.Tree.In<A: Type>(tree: PriorityQueue.Tree<A>, a: A): Type
  case tree {
    empty:
      Empty
    node: 
      PriorityQueue.Either<
        PriorityQueue.Tree.In(tree.left),
        a == tree.here,
        PriorityQueue.Tree.In(tree.right)
      >
  }
```

**Reply by dvberkel** [reply]
Where `PriorityQueue.Either` and `PriorityQueue.Tree.In.aux` should be the same :thinking:

**Reply by rigille** [reply]
oops let me fix this

**Comment by dvberkel** [comment]
Progress on the @rigille suggestions

* [x] `PriorityQueue.In`
* [x] `PriorityQueue.priority`
* [x] `PriorityQueue.safe_head`
* [x] `PriorityQueue.In.not_empty`
* [ ] `PriorityQueue.head_ok`
